review Test 초안입니다.

sngsngUDON opened the pull request:
## #️⃣연관된 이슈

> #31 

## 📝작업 내용

> Review 기능에 관련된 Test 초안 작성했습니다. 기능 더 구현하면 test 더 업데이트 하겠습니다.

### 스크린샷 (선택)


## 💬리뷰 요구사항(선택)

>

sngsngUDON committed:
review Test

sngsngUDON committed:
Store.jsp js/css/jsp 파일로 분리 및 store 기능 부분 개선

1. Store.jsp 에 있었던 Javascript 부분과 css 부분을 파일 분리했습니다.
2. Store 부분의 부족했던 기능을 개선

sngsngUDON committed:
md파일 삭제(따로 기록)

sngsngUDON committed:
들여쓰기 수정 및 BEAN 주입 에러 해결

5y1ee commented on src/main/java/kr/co/solfood/user/review/ReviewController.java:
수정된 내용 좋습니다. 경로도 /user/store 에 이의는 없습니다만 자기가 만든 컨트롤러 및 api들 명세 작성은 미리 해주세요.
/user/store : 상점 메인 페이지
/user/store?storeId= : 특정 상점 상세 페이지

이런 식으로.

5y1ee replied:
ReviewController인데 매핑이 /user로 되어있네요. 이 부분 이전에 언급했을텐데, StoreController 참고해서 수정해주세요.
그리고 StoreController에 추가된 부분도 같이 지워주세요. (#33 참고)

5y1ee replied:
> ReviewController인데 매핑이 /user로 되어있네요. 이 부분 이전에 언급했을텐데, StoreController 참고해서 수정해주세요. 그리고 StoreController에 추가된 부분도 같이 지워주세요. (#33 참고)

StoreController 밑에 수정된 것 같네요. 무시하셔도 될 것 같습니다.

5y1ee commented on src/main/java/kr/co/solfood/user/review/ReviewController.java:
```diff
@@ -180,8 +185,33 @@ public String reviewSearch(@RequestParam String reviewTitle, Model model) {
     // 리뷰 메인 페이지
     @GetMapping("/main")
-    public String reviewMain(Model model) {
+    public String reviewMain(@RequestParam(value = "storeId", required = false) Integer storeId, Model model) {
```
이러면 리뷰 메인 페이지 주소는
/user/main 이에요? 맞다면 수정해주세요. 의미가 맞지 않습니다.

sngsngUDON replied:
지금 현재로는 그런데 리뷰 메인 페이지는 없어질 페이지라 다른 페이지 부분에서 매핑 주소좀 고치겠습니다.

5y1ee commented on src/main/java/kr/co/solfood/user/store/StoreConstants.java:
이런 클래스 좋습니다. 어쩌다가 이거 만들게 된건가요? 제가 비슷한 리뷰를 한 것 같긴한데 그게 민석님 코드였는지 기억은 안나네요.

sngsngUDON replied:
하드코딩 된 부분들이나 성능 적으로 저하가 심해 보이는 부분들을 개선하려고 찾아보다가 상수들은 저렇게 따로 모아 관리하면 좋다고 들어서 저런 식으로 분리했습니다.

5y1ee commented on src/main/java/kr/co/solfood/user/store/StoreController.java:
```diff
@@ -31,39 +41,89 @@ public String getStoreList(@RequestParam(value = "category", required = false) S
-            e.printStackTrace();
-            return ResponseEntity.badRequest().build();
-        }
+        Map<String, Object> config = categoryProperties.getCategoryConfig();
```
이 페이지의 존재 의의는?

5y1ee replied:
store.js 에서 쓰는 api 임. ok

여기서 리뷰를 추가하자면, 사용자의 직접 접근을 막을 순 없나? 결국 프로퍼티 데이터를 가져오는 건데 이런 방식이 최선인가? 에 대한 의문.

sngsngUDON replied:
Review 작성, 삭제, 수정도 User 로그인 한 상태에서만 할 수 있도록 인터셉터에 추가했고, 이 부분도 AdminHomeController.java 파일로 옮겼습니다.

5y1ee replied:
인터셉터 얘기가 아니고, 로그인을 한 유저가 실제로 url에 api/store... 뭐 이런 식으로 입력을 하면 해당 api 호출이 가능하지 않을까요? 그걸 막을 방법은 없는가, 에 대한 이야기 입니다.

sngsngUDON replied:
이 부분은 검색 기능에서 다시 storeController 에 있어야 사용하기 편한거 같아서 다시 끌어오겠습니다. 말하신 부분은 조금 더 고민해볼게요

5y1ee commented on src/main/java/kr/co/solfood/user/store/StoreController.java:
```diff
@@ -31,39 +41,89 @@ public String getStoreList(@RequestParam(value = "category", required = false) S
 
+    // 상점 상세 페이지로 이동
+    @GetMapping("/detail/{storeId}")
+    public String getStoreDetail(@PathVariable int storeId) {
```
이 함수 접근 방법이 어떻게 되죠?
혹시 storeId를 검증할 필요는 없을까요? 존재하지 않는 페이지로 이동 요청이 오면 그냥 404로 처리한다 라고 쳐도 되긴 합니다.

sngsngUDON replied:
user/store/detail/ [ storeId ] 로 storeId 에 맞는 가게의 상세페이지로 이동합니다.

5y1ee replied:
매핑 주소를 묻는건 아니고, storeId의 validation을 체크할 필요가 없는가 에 대한 논의입니다.
특정 가게를 클릭하면 해당 가게의 id를 받아와서 user/store/detail/ [ storeId ] 로 이동을 시키는 구조인가요?

5y1ee commented on src/main/java/kr/co/solfood/user/store/StoreController.java:
```diff
@@ -31,39 +41,89 @@ public String getStoreList(@RequestParam(value = "category", required = false) S
+     */
+    @ExceptionHandler(StoreException.class)
+    @ResponseBody
+    public ResponseEntity<Map<String, Object>> handleStoreException(StoreException e) {
```
좋은 시도 같습니다.

5y1ee commented on src/main/resources/kr/co/solfood/user/store/StoreMapper.xml:
```diff
@@ -46,8 +46,13 @@
-        ORDER BY store_id DESC
+        <where>
+            <if test="category != null and category != '' and category != '전체'">
+                AND store_category LIKE CONCAT(#{category}, '%')
```
concat 문이 이전과 달라졌는데 이유가 있을까요?

sngsngUDON replied:
제가 성능 향상 방법을 요청했는데 원래 하던 방식이 매우 비효율적이고, 수정한 방법이 인덱스를 활용할 수 있는 성능 면에서도 좋은 방법이라고 받았습니다.

5y1ee commented on src/main/webapp/js/store.js:
```diff
@@ -0,0 +1,395 @@
+    }
+}
+
+function placesSearchCB(data, status, pagination) {
```
CB는 CallBack이다..

5y1ee commented on src/test/java/kr/co/solfood/user/review/ReviewServiceImplTest.java:
```diff
@@ -0,0 +1,238 @@
+        Double result = reviewService.getAverageStar();
+        
+        assertNotNull(result);
+        assertEquals(4.5, result);
```
이 테스트는 거의 실패할 것 같은데 의도가 있나요?

5y1ee commented on src/test/java/kr/co/solfood/user/review/ReviewServiceImplTest.java:
```diff
@@ -0,0 +1,238 @@
+        Integer result = reviewService.getTotalCount();
+        
+        assertNotNull(result);
+        assertEquals(20, result);
```
이것도.. 위랑 같음

5y1ee commented on src/test/java/kr/co/solfood/user/review/ReviewServiceImplTest.java:
```diff
@@ -0,0 +1,238 @@
+     */
+    private ReviewVO createTestReview() {
+        ReviewVO review = new ReviewVO();
+        review.setReviewId(1);
```
reviewId는 pk로 autoincrement로 만드는거 아닌지?

sngsngUDON replied:
네 맞네요 그 정보를 따로 입력을 안해서 저렇게 나온거 같습니다

5y1ee commented on src/test/java/kr/co/solfood/user/review/ReviewServiceImplTest.java:
```diff
@@ -0,0 +1,238 @@
+    /**
+     * ReviewMapper의 가짜 구현체 - 실제 DB 없이 테스트 데이터 반환
+     */
+    static class TestReviewMapper implements ReviewMapper {
```
굳이 이 클래스가 필요한지,
우린 db 연결해서 테스트하거나 mock으로 db 연결없이 되지 않을까 싶은데... 잘 모르겠다;;

sngsngUDON replied:
지금 한거는 only junit 으로만 테스트 코드를 짜봤는데 아직 미흡한 부분이 많은 것 같아서 위에 부분들과 같이 개선해보겠습니다.

5y1ee reviewed (commented):
user로 올리는건데 reviewer를 4명 다 설정한 이유는? 연관작업자 한명만 등록해주세요. develop이면 저 등록해주시고.

테스트 코드가 잘못된건 아닌데 우리가 하던 방식이랑 많이 다르네요. 그래서 판단이 좀 힘듭니다. 무엇보다, 본인 코드에 대한 인지를 더 하고 올려주세요.

sngsngUDON committed:
인터셉터 파일 수정 및 에러 발생시 페이지 이동 통일

1. User 로그인이 됐을 때만 리뷰 작성, 수정, 삭제를 할 수 있도록 인터셉터 파일 변경했습니다.
2. 카테고리 설정 정보 API 를 AdminHomeController 로 이동했습니다.
3. review 관련 url 매핑을 수정했습니다.

sngsngUDON reviewed (commented):
리뷰 수정했습니다.

5y1ee commented on src/main/java/configuration/MvcConfig.java:
```diff
@@ -166,7 +166,11 @@ public void addInterceptors(InterceptorRegistry registry) {
+                .excludePathPatterns("/user/store/**")          // 상점 목록/상세는 로그인 없이 접근 가능
+                .excludePathPatterns("/user/review/list")       // 리뷰 목록은 로그인 없이 접근 가능
+                .excludePathPatterns("/user/review/detail/**")  // 리뷰 상세보기는 로그인 없이 접근 가능
+                .excludePathPatterns("/user/review/search");    // 리뷰 검색은 로그인 없이 접근 가능
```
만약 이렇게 인터셉터를 처리하면, 제외 설정한 모든 페이지에 대해 회원/비회원 기능 2가지 모두 구현해야 할 겁니다.
그 점도 염두해두세요.

sngsngUDON replied:
네 확인했습니다.

5y1ee commented on src/main/java/kr/co/solfood/admin/home/AdminHomeController.java:
```diff
@@ -36,5 +45,18 @@ public String userSearch(@RequestParam String query, Model model) {
+    //카테고리 설정 정보 API (관리자용)
+    @GetMapping("/api/category/config")
+    @ResponseBody
+    public ResponseEntity<Map<String, Object>> getCategoryConfig() {
```
이걸 admin으로 옮기라고 얘기한게 아니라, user든 admin 이든 사용자가 직접 api/category/... 이 주소로 이동을 시도했을 때 뜨면 안될 것 같단 의미입니다.

5y1ee commented on src/main/java/kr/co/solfood/user/review/ReviewController.java:
```diff
@@ -35,14 +35,14 @@ public class ReviewController {
-            // 가게 정보가 없으면 store 메인 페이지로 리다이렉트
-            return "redirect:/user/store";
+            // 존재하지 않는 가게면 404 페이지 표시
+            return "error/404";
```
음? 이거는 메인 페이지로 리다이렉트 해도 좋습니다. 오히려 404보다 메인 페이지로 그냥 보내는게 자연스러울 수도 있을 것 같아요

5y1ee replied:
이 부분 말고 다른 404 return 하는 부분 모두 포함해서 말한겁니다.

sngsngUDON committed:
파일명 변경, url 매핑 변경

1. reviewList.jsp 를 review 폴더 아래 list.jsp 로 변경했습니다. (중복이슈)
2. 카테고리 설정 정보 API 를 다시 storeController 로 옮겼습니다. (검색어 할 때 사용이 됩니다.)
3. 각 에러 상황에 맞게 url 매핑을 개선했습니다.

sngsngUDON committed:
Merge branch 'user' into minseok/user/review

sngsngUDON committed:
MvcConfig파일 수정
